core/reflection.jvm/src/kotlin/reflect/jvm/internal/KProperty0Impl.kt

```diff
@@ -42,6 +42,13 @@ internal open class KProperty0Impl<out V> : KProperty0<V>, KPropertyImpl<V> {
 
     class Getter<out R>(override val property: KProperty0Impl<R>) : KPropertyImpl.Getter<R>(), KProperty0.Getter<R> {
         override fun invoke(): R = property.get()
+
+        override fun equals(other: Any?): Boolean {
+            val that = other as? Getter<*> ?: return false
```

Comment:
```property.name``` is not sufficient as a basis for equality, and it is not feasible to use ```property==that.property```  directly because it requires equality of ```rawBoundReceiver```.
So ```property.toString() == that.property.toString()``` seems good to me.

Reply:
Receiver should have effect on `equals`, can you clarify why do you think it's not feasible?

`equals` is implemented in this way for properties already, so I think it should be consistent for property accessors too. E.g. in the test above, `cProp` and `C_prop` are actually not equal and that is by design.

Reply:
I agree that ```Receiver``` should have an impact on the equality. I thought it is not feasible because:
1.  It causes  ```cProp.setter != C()::prop.setter```, and that seems because those two instance of ```C``` are not equal
2.  maybe we can somehow make these two instance equal, but it seems unnecessary because ```KProperty0``` has no receiver and ```KProperty1``` has one receiver and we can check it by ```val that = other as? Getter<*> ?: return false```

Reply:
I mean,  if I understand correct, through the following code, receiver already has effect on ```euals```

```Kotlin
override fun equals(other: Any?): Boolean {
    val that = other as? Getter<*> ?: return false // this can  checks whether the Kproperty has one receiver
    return property.toString() == that.property.toString() // this can check the receiver name
}
```

Reply:
Sorry for the delay. By "receiver" above I meant the value to which the property is bound to, or in other words, the left-hand side of the `::` expression.

Currently, if receivers are different, `KProperty` instances are different too. For example, `C()::prop == C()::prop` will be **false** and that is correct. `KProperty.Getter/Setter` instances should follow the same logic, and be equal only in case all of the components of the property (including the bound receiver) are equal.

From this it follows that the implementation should basically be:
```
override fun equals(other: Any?): Boolean =
    other is Getter<*> && property == other.property
```

It seems that the commented lines in the test in master (https://github.com/JetBrains/kotlin/blob/master/compiler/testData/codegen/box/callableReference/bound/equals/propertyAccessors.kt#L33..L34) are actually incorrect, and thus should be changed to reflect the more intuitive behavior.

Reply:
Very detailed and clear,  thank you.